src/Abp/Runtime/Caching/AbpCacheBase.cs

```diff
@@ -77,51 +77,51 @@ public virtual TValue Get(TKey key, Func<TKey, TValue> factory)
                 }
             }
 
-            return item;
+            return value;
         }
 
         public virtual TValue[] Get(TKey[] keys, Func<TKey, TValue> factory)
         {
-            TValue[] items = null;
+            TValue[] values = null;
 
             try
             {
-                items = GetOrDefault(keys);
+                values = GetOrDefault(keys);
             }
             catch (Exception ex)
             {
                 Logger.Error(ex.ToString(), ex);
             }
 
-            if (items == null)
+            if (values == null)
             {
-                items = new TValue[keys.Length];
+                values = new TValue[keys.Length];
             }
 
-            if (items.Any(i => i == null))
+            if (values.Any(i => i == null))
             {
                 lock (SyncObj)
                 {
                     try
                     {
-                        items = GetOrDefault(keys);
+                        values = GetOrDefault(keys);
                     }
                     catch (Exception ex)
                     {
                         Logger.Error(ex.ToString(), ex);
                     }
 
                     var fetched = new List<KeyValuePair<TKey, TValue>>();
-                    for (var i = 0; i < items.Length; i++)
+                    for (var i = 0; i < values.Length; i++)
                     {
                         TKey key = keys[i];
-                        TValue value = items[i];
-                        if (value == null)
+                        TValue value = default;
+                        if (EqualityComparer<TValue>.Default.Equals(values[i], default))
                         {
-                            value = factory(key);
+                            values[i] = value = factory(key);
                         }
 
-                        if (value != null)
+                        if (!EqualityComparer<TValue>.Default.Equals(value, default))
```

Comment:
This means a default value returned by the factory can never be cached.

Reply:
i think this is acceptable. given that only default values of primitive types and struct will not be cache.

Reply:
Why acceptable? It is an unnecessary bug.

Reply:
This isn't a bug, currently it works the same way without the changes in this PR.

Looking at the current implementation, assuming TValue is int (non-nullable).
The `item` returns by `item = GetOrDefault(key)` will never be `null`, therefore, the code path will never enter `if (item == null)` (which calls `Set()`)
https://github.com/aspnetboilerplate/aspnetboilerplate/blob/629ca574e6d086400438f06f9bd1e7854840a985/src/Abp/Runtime/Caching/AbpCacheBase.cs#L35-L46

Reply:
An example for `0` would be caching a count of objects where the actual count is `0`.
It is also reasonable to want to cache the default value after the `factory` is run once.

> ~Looking at the current implementation, assuming TValue is int (non-nullable).
The `item` returns by `item = GetOrDefault(key)` will never be `null`, therefore, the code path will never enter `if (item == null)` (which calls `Set()`)~

Indeed, this seems to be the case ⁠— why did you strike it out?
That is, even if we cache a default value, the current implementation won't be able to tell.

> ~This isn't a bug, currently it works the same way without the changes in this PR.~

This is a bug introduced in this PR for the `0` case, but it already exists for the `null` case.

Reply:
Yup I understand that we probably should allow caching of 0 for int, for example.

That’s why I stroke out my comment and implemented the change to allow default value to be cached. 😅

Reply:
Continued in https://github.com/aspnetboilerplate/aspnetboilerplate/pull/5277#discussion_r380255395.